### PR TITLE
Allow the Consumer to disable auto-commit offsets

### PR DESCRIPTION
**Pull request by kjelle**
Adding configuration `Consumer.Offsets.Enable` (default `true`)
It blocks the call to `flushToBroker` in `OffsetManager`

### CODE REVIEW

**Comment by kjelle** [comment]
One of the tests fail;

> coverage:  71.3% of statements
> panic: test timed out after 1m30s

How does one force a rerun @eapache  ?

**Comment by snigle** [comment]
Hi, I also need to disable the autocommit but I need to commit when I want and at closing.
Do you have the same need ?
If yes this patch should be enough ?  https://github.com/Shopify/sarama/pull/1193

**Comment by kjelle** [comment]
Given the new consumer groups one also needs to escape in the Close() of the offset manager as it requires MainLoop to run... 
I updated my patch @snigle

**Comment by kjelle** [comment]
@taiyang-li I'll merge upstream and see if I is compatible with the latest tag, 1.23.1 - if so, I'll reissue this PR.

**Comment by kjelle** [comment]
cla is signed. im unsure why only one of the travis builds fail, and not all..

**Comment by varun06** [comment]
hey @kjelle 
even though the change is simple, I feel that there should be a test for it, don't want an edge case to pop up.

is it feasible to write a test for it?

File: config.go [hunk 1]
```diff
@@ -423,6 +427,7 @@ func NewConfig() *Config {
 	c.Consumer.MaxWaitTime = 250 * time.Millisecond
 	c.Consumer.MaxProcessingTime = 100 * time.Millisecond
 	c.Consumer.Return.Errors = false
+	c.Consumer.Offsets.Enable = true
```
Comment (sam-obeid) [inline config.go]: Would suggest have a more descriptive name:

```suggestion
	c.Consumer.Offsets.AutoCommitEnable = true
```

Also `c.Consumer.Offsets.AutoCommit.Enable` would be fine.

Comment (varun06) [reply config.go]: @kjelle can you please make this change?

Comment (kjelle) [reply config.go]: > Also `c.Consumer.Offsets.AutoCommit.Enable` would be fine.

I totally agree @sam-obeid. I am making appropriate changes and reissuing the PR.

**Comment by kjelle** [comment]
> is it feasible to write a test for it?

I am adding tests to check if setting the `Consumer.Offsets.AutoCommit.Enable` gives the expected results.

**Comment by kjelle** [comment]
@varun06 I discovered some problems while writing the unit-tests. I am rewriting how the offset_manager.go will handle a disabled `Consumer.Offsets.AutoCommit.Enable`. I will test it in our production line before re-submitting the PR.

**Comment by varun06** [comment]
@kjelle That's why I asked for test :)

Thanks for doing that and finding the issue.

**Comment by varun06** [comment]
@sam-obeid can you please have a look again?

**Comment by kjelle** [comment]
@varun06 given the amount of commits I would suggest a merge squash when it comes to that :)

**Comment by varun06** [comment]
Done, Please test and let me know how it goes.

File: config.go [hunk 2]
```diff
@@ -338,8 +338,15 @@ type Config struct {
 		// offsets. This currently requires the manual use of an OffsetManager
 		// but will eventually be automated.
 		Offsets struct {
-			// How frequently to commit updated offsets. Defaults to 1s.
-			CommitInterval time.Duration
```
Comment (d1egoaz) [inline config.go]: ohh shit, this was a breaking change, i.e: it breaks people using sarama-cluster with the latest version of sarama.
```
../../../../go/pkg/mod/github.com/bsm/sarama-cluster@v2.1.15+incompatible/consumer.go:452:59: c.client.config.Config.Consumer.Offsets.CommitInterval undefined (type struct { AutoCommit struct { Enable bool; Interval time.Duration }; Initial int64; Retention time.Duration; Retry struct { Max int } } has no field or method CommitInterval)
```

File: offset_manager.go [hunk 3]
```diff
@@ -233,7 +233,12 @@ func (om *offsetManager) mainLoop() {
 	}
 }
 
+// flushToBroker is ignored if auto-commit offsets is disabled
 func (om *offsetManager) flushToBroker() {
+	if !om.conf.Consumer.Offsets.AutoCommit.Enable {
```
Comment (imaffe) [inline offset_manager.go:238]: Glad this line is fixed, it actually blocks all calls to Commit(), however sometimes we do wish to commit manually.

**Comment by kjelle** [comment]
I'm sorry if this PR caused a lot of hazzle. I commit offsets manually by making my own offset manager sending a sarama.OffsetCommitRequest with my own block containing topic, partition and offset. I then send it to the correct broker coordinating that partition. If Sarama now supports manual offset commit I could move to that instead.
